### ISSUE TITLE
Async routes dies because of unowned self

### ISSUE THREAD

**Issue by fizker** [IBM-Swift/Kitura#563]
A route that needs to do async work (such as talk to another server via `NSURLSession`) fails. I discovered this, because I was trying to have a Kitura-based server talk to the Apple Push Notification Service.

It dies in https://github.com/IBM-Swift/Kitura/blob/0.19.2/Sources/Kitura/RouterMiddlewareWalker.swift#L57 because `self` is unowned and have been killed at time of execution.

This happens every time.

Minimum example code (in Kitura 0.19.2):
```swift
import Foundation
import Kitura
let router = Router()
router.get("/async") { request, response, next in
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)) {
		response.send("some result")
		next() // this line explodes
	}
}
Kitura.addHTTPServer(onPort: 8090, with: router)
Kitura.run()
```

**Comment by simplepanda** [comment]
You don't need to dispatch_async() to do work in your router function. The function itself is executed via dispatch by the Router(), so any long lived work you do in your function delays only the response to the client and not other requests to Kitsura (even other requests to the same route). 

You can talk to APNS / your database / etc in your function and just respond when your function is complete without sub-dispatching anything.

**Comment by fizker** [comment]
I didn't do `dispatch_async()`, it was just the simplest way to reproduce the issue. What I did was rougly

```swift
let task = session.dataTask(with:url) { data, _, _ in
    // do something with task
    response.send(result)
    next()
}
task.resume()
```

And if the route handlers are meant to be synchronous, why do you pass in a `next` func? Simply returning from the handler would achieve the same result. A `next` func only makes sense for doing async work.

**Comment by ianpartridge** [comment]
@shmuelk ?

**Comment by simplepanda** [comment]
I'm supposed to be heading to the office right nows so I just quickly glanced back through -Sys, -Net, and Kitura proper but based on a speed read...

response.send() in your router function doesn't send any data - it just adds your data to a buffer on the Kitura side. The next() call is what actually sends the headers, flushes the buffer to the socket, etc. It's not related to async dispatching in any way.

You can see where the next() function code lives in Kitura/Router.swift - 168-180 (including the fact that, as you say, it's unowned self):

`process(request: routeReq, response: routeResp) { [unowned self] () in
            do {
                if  !routeResp.state.invokedEnd {
                    if  routeResp.statusCode == .unknown  && !routeResp.state.invokedSend {
                        self.sendDefaultResponse(request: routeReq, response: routeResp)
                    }
                    try routeResp.end()
                }
            } catch {
                // Not much to do here
                Log.error("Failed to send response to the client")
            }
        }
`

**Comment by fizker** [comment]
Ahh. I have done a fair bit of nodejs development, and in that world, a callback called `next` or `done` means async handling.

It then seems that the solution in my case would be to create an explicit `NSOperationQueue`, pass it to the `NSURLSession` and then call `queue.waitUntilAllOperationsAreFinished()` before returning from the handler. Have I understood that correctly?

**Comment by dfirsht** [comment]
Well all Kiutra middleware is handling asynchronously so there should be no reason why you cannot call next inside your callback. As to what @simplepanda was saying, to send data you can either call `next()` and data will eventually be sent at the end of the pipeline; or you can call `end()` which will terminate the pipeline and send the data to the client right away.

Either way though it shouldn’t make a difference in the error you are seeing. Can you please post the error message?

**Comment by fizker** [comment]
Assertion failed: (object->weakRefCount.getCount() && "object is not currently weakly retained"), function swift_unownedRetainStrongAndRelease, file /Users/buildnode/jenkins/workspace/oss-swift-package-osx/swift/stdlib/public/runtime/HeapObject.cpp, line 516.
![screen shot 2016-06-27 at 16 33 02](https://cloud.githubusercontent.com/assets/736545/16383314/e4bb71cc-3c84-11e6-86e6-f7c8f8f3e50c.png)

The snippet in the OP produces this result every time.

**Comment by fizker** [comment]
I just tried replacing the `next()` call with `try! response.end()`, and then it works as expected. I guess I can live with that workaround for now :).

**Comment by dfirsht** [comment]
Alright well we can look into why your first snippet fails anyway. Btw, you can just do `try response.end()` and the framework will handle any errors thrown.

**Comment by fizker** [comment]
`try response.end()` does not compile inside the `block` passed to `dispatch_async`, as the example is doing. But otherwise you are correct :)

**Comment by dfirsht** [comment]
@fizker, I investigated this issue and found that we were incorrect in adding unowned self to our walkers. I fixed this issue in release 0.19.4. You should be able to make asynchronous calls in your callbacks now.

**Comment by fizker** [comment]
@dfirsht It does indeed work in 0.19.4. I'll close the issue then
